runtime/sema/positioninfo.go

```diff
@@ -0,0 +1,251 @@
+/*
+ * Cadence - The resource-oriented smart contract programming language
+ *
+ * Copyright 2019-2022 Dapper Labs, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package sema
+
+import (
+	"github.com/onflow/cadence/runtime/ast"
+	"github.com/onflow/cadence/runtime/common"
+	"math"
+)
+
+type PositionInfo struct {
+	Occurrences         *Occurrences
+	VariableOrigins     map[*Variable]*Origin
+	MemberOrigins       map[Type]map[string]*Origin
+	MemberAccesses      *MemberAccesses
+	Ranges              *Ranges
+	FunctionInvocations *FunctionInvocations
+}
+
+func NewPositionInfo() *PositionInfo {
+	return &PositionInfo{
+		MemberOrigins:       map[Type]map[string]*Origin{},
+		VariableOrigins:     map[*Variable]*Origin{},
+		Occurrences:         NewOccurrences(),
+		MemberAccesses:      NewMemberAccesses(),
+		Ranges:              NewRanges(),
+		FunctionInvocations: NewFunctionInvocations(),
+	}
+}
+
+func (i *PositionInfo) recordVariableReferenceOccurrence(
+	memoryGauge common.MemoryGauge,
+	startPos ast.Position,
+	endPos ast.Position,
+	variable *Variable,
+) {
+	origin, ok := i.VariableOrigins[variable]
+	if !ok {
+		startPos2 := variable.Pos
+		var endPos2 *ast.Position
```

Comment:
maybe rename `startPos2` and `endPos2` to `originStartPos` and `originEndPos`

Reply:
Good idea! Improved in e0a663d0b66d89348666aab002126a44dfdacdf7